SparkPost/2web2ui
# FE-1135 - Incorporate Cypress tests in to code coverage calculation

[FE-1135](https://jira.int.messagesystems.com/browse/FE-1135)

### What Changed
- Updates Cypress to `4.11.0`
- Adds `@cypress/code-coverage` plugin to help calculate code coverage
- Adds `cypress-test-retries` to help with flaky tests - though this will be incorporated by the Cypress team as a core feature shortly - https://docs.cypress.io/guides/references/roadmap.html#Upcoming-features
- Adds common configuration options for Cypress to `cypress/plugins/index.js`
- Adds `cypress.dev.json` and `cypress.ci.json` configuration files that have some additional configuration for values found within `cypress/plugins/index.js`
- Revamps CircleCI pipeline to use a production build of the app vs. a development build
- Instruments code coverage within production builds when the env `INSTRUMENT_CODE=true` is passed - see `npm run build-ci`
- Uses `nyc` to merge code coverage reports from Jest (4 runners) and Cypress (10 runners) before uploading a single, merged report to Codecov

### How To Test
- For production builds, verify that _no code instrumentation is present_ - in the console, type `window.__coverage__` => should return `undefined`
- Verify tests continue to pass as expected
- Go to https://codecov.io/gh/SparkPost/2web2ui and verify this PR has a code coverage report

### To Do
- [x] Figure out a good way to instrument code for a production build
- [x] Re-incorporate `@cypress/code-coverage`
- [x] Re-incorporate Hibana Cypress tests
- [x] Promote build artifacts to next job so that build step does not need to run each time
- [x] Properly blacklist API requests for pseudo staging environment
- [x] is `nyc` a necessary dependency? - nope!
- [x] Verify that coverage is negatively impacted when Cypress tests are removed
- [x] Verify that coverage is negatively impacted when Jest tests are removed
- [x] Codecov upload is silently failing sometimes, yet exiting with code 0
